cql3: Represent create_statement using managed_bytes

dawmd opened the pull request:
When describing a table, we need to do it carefully: if some
columns were dropped, we must specify that explicitly by

```
ALTER TABLE {table} DROP {column} USING TIMESTAMP ...
```

in the result of the DESCRIBE statement. Failing to do so
could lead to data resurrection.

However, if a table has been altered many, many times,
we might end up with a huge create statement. Constructing
it could, in turn, trigger an oversized allocation.
Some tests ran into that very problem in fact.

In this commit, we want to mitigate the problem: instead of
allocating a contiguous chunk of memory for the create
statement, we use `bytes_ostream` and `managed_bytes` to
possibly keep data scattered in memory. It makes handling
`cql3::description` less convenient in the code, but since
the struct is pretty much immediately serialized after
creating it, it's a very good trade-off.

A reproducer is intentionally not provided by this commit:
it's easy to test the change, but adding and dropping
a huge number of columns would take a really long amount
of time, so we need to omit it.

Fixes scylladb/scylladb#24018

Backport: all of the supported versions are affected, so we want to backport the changes there.

piodul commented:
> A reproducer is intentionally not provided by this commit:
> it's easy to test the change, but adding and dropping
> a huge number of columns would take a really long amount
> of time, so we need to omit it.

Did you try it? How slow was it?

I think you can make the test faster if you use long column names and complicated types with long descriptions.

denesb commented on cql3/description.cc:
```diff
@@ -18,21 +18,21 @@ static logging::logger dlogger{"description"};
 
 namespace cql3 {
```
does this commit compile on its own?

denesb commented on schema/schema.cc:
```diff
@@ -1162,12 +1164,12 @@ std::ostream& schema::schema_properties(const schema_describe_helper& helper, st
-    os << "\n    AND max_index_interval = " << max_index_interval();
-    os << "\n    AND memtable_flush_period_in_ms = " << memtable_flush_period();
-    os << "\n    AND min_index_interval = " << min_index_interval();
+    os << "\n    AND crc_check_chance = " << std::format("{}", crc_check_chance());
```
fmt::to_string()

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17256/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17256/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17256/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cql_query_test
:x: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8360/)
:x: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16442/)
:x: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5192/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17256/)
#### Failed Tests (10/45165):
- [test_describe](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17256/testReport/junit/cqlsh_tests.cqlsh_tests/TestCqlsh/Tests___dtest_with_consistent_topology_changes___test_describe) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_describe)
- [test_describe[require_client_auth=true]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17256/testReport/junit/cqlsh_tests.cqlsh_tests/TestCqlshWithSSL/Tests___dtest_with_consistent_topology_changes___test_describe_require_client_auth_true_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_describe[require_client_auth=true])
- [test_describe[require_client_auth=false]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17256/testReport/junit/cqlsh_tests.cqlsh_tests/TestCqlshWithSSL/Tests___dtest_with_consistent_topology_changes___test_describe_require_client_auth_false_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_describe[require_client_auth=false])
- [test_describe[require_client_auth=false]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17256/testReport/junit/cqlsh_tests.cqlsh_tests/TestCqlshWithSSL/Tests___dtest_with_gossip_topology_changes___test_describe_require_client_auth_false_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_describe[require_client_auth=false])
- [test_describe](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17256/testReport/junit/cqlsh_tests.cqlsh_tests/TestCqlsh/Tests___dtest_with_gossip_topology_changes___test_describe) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_describe)
- [test_describe[require_client_auth=true]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17256/testReport/junit/cqlsh_tests.cqlsh_tests/TestCqlshWithSSL/Tests___dtest_with_gossip_topology_changes___test_describe_require_client_auth_true_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_describe[require_client_auth=true])
- [test_describe[require_client_auth=true]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17256/testReport/junit/cqlsh_tests.cqlsh_tests/TestCqlshWithSSL/Tests___dtest_with_tablets___test_describe_require_client_auth_true_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_describe[require_client_auth=true])
- [test_describe](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17256/testReport/junit/cqlsh_tests.cqlsh_tests/TestCqlsh/Tests___dtest_with_tablets___test_describe) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_describe)
- [test_describe[require_client_auth=false]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17256/testReport/junit/cqlsh_tests.cqlsh_tests/TestCqlshWithSSL/Tests___dtest_with_tablets___test_describe_require_client_auth_false_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_describe[require_client_auth=false])
- [test_desc_restore.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17256/testReport/junit/cqlpy/test_describe/Tests___Unit_Tests___test_desc_restore_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_desc_restore.debug.1)
#### Build Details:
- Duration: 8 hr 13 min
- Builder: spider7.cloudius-systems.com

dawmd commented:
> > A reproducer is intentionally not provided by this commit:
> > it's easy to test the change, but adding and dropping
> > a huge number of columns would take a really long amount
> > of time, so we need to omit it.
> 
> Did you try it? How slow was it?
> 
> I think you can make the test faster if you use long column names and complicated types with long descriptions.

I did and it took at least 10 minutes to finish. The test looked more or less like this:

1. Create a keyspace.
2. Create a table with one column.
3. Concurrently create and drop 25,000 new columns.
4. Perform a describe.

FWIW I set the `max_schema_agreement_wait` to 0 in the test:
```py
assert hasattr(cql.cluster, "max_schema_agreement_wait")
cql.cluster.max_schema_agreement_wait = 0
```

dawmd commented:
Version 2:
* `std::format` -> `fmt::to_string`.
* Provided a reproducer.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17307/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17307/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17307/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cql_query_test
:small_blue_diamond: cluster/test_describe
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5239/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16493/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8410/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17307/)
#### Build Details:
- Duration: 7 hr 41 min
- Builder: spider7.cloudius-systems.com

piodul commented:
@michoecho please review the changes to the `bytes_ostream` and its use in this PR

michoecho commented on bytes_ostream.hh:
```diff
@@ -268,6 +268,17 @@ public:
         write(bytes_view(reinterpret_cast<const signed char*>(ptr), size));
     }
 
+    [[gnu::always_inline]]
+    void write(std::string_view sv) {
+        write(sv.data(), sv.size());
+    }
+
```
Anyway, it's not a problem to have a `void write(std::string_view sv)`.
Text spans are a subset of byte spans, after all.

The problem comes later, when we turn the output of `bytes_ostream` back into text.
This way we lose a layer of bug protection that the type system gives us.

After we start using `bytes_ostream` to produce text, we introduce the risk of a bug
where we accidentally print non-text bytes into a string, and then bad things follow.

For this use case, I would consider adding text-typed wrappers over `bytes_ostream`,
`managed_bytes` and `managed_bytes_view` which guarantee that their context
are text. (I.e. the only thing you can write to them is `std::string_view`,
and you can convert their contents to `std::string_view` without casting them yourself).

michoecho replied:
Also, just stuffing text into `managed_bytes` degrades the ergonomics of logging and debugging. Loggers and debuggers won't know that this particular `managed_bytes` is text unless you tell them explicitly.
With a typed wrapper, they can understand that this is text, and print it in a human-readable form.

michoecho reviewed (changes requested):
I'm marking this review as `request changes` because I think the order of commits 2 and 3 is wrong, see the relevant comment.

The other comments are subjective. I don't like that we just stuff text into `managed_bytes`. I think it's going to set a bad precedent. It adds one possible mode of creating bugs, and confuses debuggers and loggers. If we want to use `bytes_ostream` as a replacement for `ostringstream`, I think it would make sense to have some thin text-guaranteed wrappers around `bytes_ostream` and `managed_bytes`.

And while I realize that nobody wants to spend too much effort on a relatively unimportant patch, the later we think about this, the harder it will be to unmake. After every such patch, some of the program-encoded knowledge about what's text and what's not will be lost. And if someone wants to add typed wrappers later, he will have to go over all the uses of `managed_bytes` / `bytes_ostream` and figure out which ones are guaranteed to be text.

avikivity commented:
It's reasonable not to test performance in a unit test. It usually leads to a flaky test.

avikivity commented on cql3/description.cc:
```diff
@@ -18,32 +18,54 @@ static logging::logger dlogger{"description"};
+        // We do this to optionally move this field. It may be quite big,
+        // so it's worth a little optimization.
+        //
+        // Note: The standard specifies that the const-qualifier of the result of the expression within
+        // `result.push_back(...)` will align with the const-qualifier of the passed `desc`, cf.
+        // https://en.cppreference.com/w/cpp/language/operator_member_access.
+        result.push_back(std::forward<Desc>(desc).create_statement);
```
Wow, such a basic thing, and I didn't know it.

avikivity replied:
Interestingly, doesn't apply to a direct member reference:

```
<source>: In member function 'A&& B::get() &&':
<source>:8:27: error: cannot bind rvalue reference of type 'A&&' to lvalue of type 'A'
    8 |     A&& get() && { return a; }
      |                           ^
```

michoecho replied:
It shouldn't be interesting.

Don't confuse `rvalue reference` with `rvalue`. Value categories are (in some sense) separate from types.

For example, inside the body of `void f(A&& a) {...}`, the expressions `a` and `a.x` are lvalues.
However, given `A&& f()`, the expressions `f()` and `f().x` are xvalues. Even though both `a` and `f()` are of type `T&&`.

michoecho replied:
> So member-name could have been listed as an xvalue for a function accepting an rvalue this, but wasn't.

If you mean that C++ could have made this choice, then yes, it could have. But then `a` would have been different from an actual `(*this).a` (a dereference expression is always an lvalue). And the implicit `dereferenced_this` would behave differently from all other rvalue-reference arguments. And it would be harder to use sanely. (Easier to get your member accidentally moved by something where you don't want it).

avikivity replied:
Ah of course. Anyway, C++ is confusing.

dawmd commented on auth/service.cc:
```diff
@@ -614,21 +616,21 @@ future<std::vector<cql3::description>> service::describe_permissions() const {
 
         co_await coroutine::maybe_yield();
     }
 
-    std::ranges::sort(result, std::less<>{}, [] (const cql3::description& desc) noexcept {
-        return std::make_tuple(std::ref(desc.name), std::ref(*desc.create_statement));
-    });
+    std::ranges::sort(result, std::less<>{}, std::mem_fn(&cql3::description::name));
```
AFAIK we only sort to have the order of the output deterministic among each type of the entities. From the perspective of the user, it shouldn't matter; only the order of the types of the entities matters.

This change makes testing a bit more annoying (we don't have a deterministic order of the statements), but that's it. We could keep the sorting, but I'm afraid we'd need to implement a comparator within the scope of this PR, and I feel like it's not worth it.

dawmd replied:
Edit: I did some more research and a comparator already exists, so we can use it. I'll do that in the next iteration.

dawmd commented on auth/service.cc:
```diff
@@ -468,12 +468,14 @@ future<std::vector<cql3::description>> service::describe_roles(bool with_hashed_
         const bool is_superuser = co_await _role_manager->is_superuser(role);
 
+        sstring create_statement = produce_create_statement(formatted_role_name, maybe_hashed_password, can_login, is_superuser);
+
```
It isn't. I was convinced we had a relatively small limit for the length of roles, service levels, etc., but it turns out that our documentation doesn't mention it, and neither does Cassandra's. I'll adjust the PR.

dawmd replied:
On second thought, I'm not sure how necessary this case is. A hashed password is a sequence of bounded length, while the rest of the string is of constant size. The only variable here is the length of the role name, which is also the case for granting permissions, service levels, etc.

While it makes sense to avoid large allocations in the code responsible for describing tables, where there's room for a tragedy (a large number of columns with long names, many dropped columns that we have to list, complicated types of UDTs, etc.), I don't feel like there's a need for that in auth/service levels: I can't imagine that users choose names as long as ~2^16 characters. At the same time, we'd obfuscate the code.

It's true that there's technically no restriction on how long those names can be (should we impose some?), but the problem is bigger—attempting to *create* a role allocates a string with its name, and we'll run into the same issue there.

I'll focus on addressing issues in UDTs/UDFs/tables/... for now, but if you think we should also cover auth/service levels, I'll do that too.

avikivity replied:
It's not necessary. I missed the context that it's for CREATE ROLE. No one will use megabyte role names.

dawmd commented on cql3/description.cc:
```diff
@@ -18,32 +18,54 @@ static logging::logger dlogger{"description"};
+namespace {
+
+template <typename Desc>
+std::vector<managed_bytes_opt> serialize_description(Desc&& desc, bool serialize_create_statement) {
```
> Does this need to be a template?

It doesn't, but the function is never exposed to any other TU, so it has no impact on compilation time, while we don't have to duplicate code. But I agree that maybe it doesn't really make sense to micro-optimise things like this that are only called once in a while.

I'm not sure if I understand what you mean by capturing the essence in an interface. Could you elaborate? One thing I can think of is you're suggesting creating a concept that identifies those types that can be described and produce a create statement—something similar to what we used to have and what was removed in this commit: scylladb/scylladb@b3573074061e1d4c945786fafa6f6ef89235d13c.

avikivity commented:
For any single statement, the only ones that are problematic are CREATE TABLE (because of deleted columns) and CREATE FUNCTION (for scalar functions). Anything else will have real-life limits. One does't create tables with tens of thousands of columns, but one can (and unfortunately does) write loops that delete and re-create columns.

dawmd commented:
Version 3:
* Rebased on top of the current master branch. (Unfortunately I didn't notice that, so the diff is bloated. Sorry for that in advance.)
* Introduced the wrappers for `managed_bytes` and `bytes_ostream`. There's intentionally no copy constructor for either of those.
* That made `cql3::description` move-only, so I adjusted the existing code to it. Fortunately, it was possible without any problems. We never copied descriptions, so it's okay. If that's too much microoptimisation for not gaining anything in return, I can change it.
* CREATE TABLE/MV/INDEX, CREATE FUNCTION and CREATE AGGREGATE use fragmented buffers. The rest of schema entities work as before, with a contiguous one.
* `cql3::description::create_statement` has `operator<` again, so we preserve the same order of statements as before this PR.

avikivity commented on utils/managed_string.hh:
```diff
@@ -0,0 +1,64 @@
+    managed_string to_managed_string() && {
+        return managed_string::from_managed_bytes_unsafe(std::move(_impl).to_managed_bytes());
+    }
+};
```
Surprised no formatter, but can be added later when needed.

avikivity replied:
Ah, it's a sink for formatting. But can't we make fmt::format use it as a sink so we don't have to use old-style formatting?

See fmt::format_to().

avikivity replied:
bytes_ostream and managed_bytes share formats, see bytes_ostream::to_managed_bytes() &&.

avikivity commented on cql3/functions/user_function.cc:
```diff
@@ -75,23 +75,33 @@ description user_function::describe(with_create_statement with_stmt) const {
-                _return_type->cql3_type_name_without_frozen(),
-                _language,
-                _body);
+        std::ostringstream os;
```
fmt::format_to() would be better here.

avikivity replied:
@dawmd ^^

michoecho reviewed (changes requested):
After looking at the points related to `managed_string` I added some suggestions for extra helpers, to avoid some back-and-forth between string views and byte views.

I think the right goal for this typing exercise is to have no casts from raw bytes into strings outside of the body of `managed_string::from_managed_bytes_unsafe`, and to have no calls to `managed_string::from_managed_bytes_unsafe` outside of `fragmented_ostringstream::to_managed_string`.

I don't insist, though. IMHO the most important part was that `description::create_statement` is made a string type (because that has some "global" consequences, and making that change later would basically require reviewing all usages of that field), and that part's done.

So if you're tired of the review ping-pong, you can ignore my suggestions. In that case I'm going to post them as a follow-up PR.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17567/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17567/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17567/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cql_query_test
:small_blue_diamond: cluster/test_describe
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16674/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8591/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5423/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17567/)
#### Build Details:
- Duration: 10 hr
- Builder: spider9.cloudius-systems.com

michoecho reviewed (approved):
Looks good to me.

avikivity commented:
I'm unhappy with the regression from fmt to std::ostream. Can't we keep fmt?

denesb commented:
Removed merge candidate label because there is an outstanding comment from Avi.

dawmd commented:
Version 5:
* Rebased on top of the current master branch.
* Describing UDFs still uses fmt instead of `operator<<`.

avikivity commented on cql3/statements/describe_statement.cc:
```diff
@@ -283,12 +284,15 @@ future<std::vector<description>> table(const data_dictionary::database& db, cons
+        os << *table_desc.create_statement;
+        os << "\n*/";
+
+        table_desc.create_statement = std::move(os).to_managed_string();
```
Forgot this one?

dawmd replied:
Yes, thank you

avikivity commented on schema/schema.cc:
```diff
@@ -1170,12 +1173,12 @@ std::ostream& schema::schema_properties(const schema_describe_helper& helper, st
+    os << "\n    AND gc_grace_seconds = " << fmt::to_string(gc_grace_seconds().count());
+    os << "\n    AND max_index_interval = " << fmt::to_string(max_index_interval());
+    os << "\n    AND memtable_flush_period_in_ms = " << fmt::to_string(memtable_flush_period());
+    os << "\n    AND min_index_interval = " << fmt::to_string(min_index_interval());
```
These seem unneeded.

dawmd replied:
They are. `fragmented_ostringstream` doesn't have an overload of `operator<<` (or `write`) that accepts all integer types: the only one present is `operator<<(char)`. If we don't convert these to to a string, we'll have a narrowing conversion.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

| Mode | Stage | Status | Node |
|------|-------|--------|------|
|  | [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18033/) | :white_check_mark: | thor |
| dev | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18033/) | :white_check_mark: | spider8.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18033/) | :white_check_mark: | spider8.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18033/)<br> :small_blue_diamond: test/boost/cql_query_test.cc<br> :small_blue_diamond: cluster/test_describe | :white_check_mark: | spider8.cloudius-systems.com |
| release | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18033/) | :white_check_mark: | spider9.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18033/) | :white_check_mark: | spider9.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18033/)<br> :small_blue_diamond: test/boost/cql_query_test.cc<br> :small_blue_diamond: cluster/test_describe | :white_check_mark: | spider9.cloudius-systems.com |
| debug | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18033/) | :white_check_mark: | spider1.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18033/) | :white_check_mark: | spider1.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18033/)<br> :small_blue_diamond: test/boost/cql_query_test.cc<br> :small_blue_diamond: cluster/test_describe | :white_check_mark: | spider1.cloudius-systems.com |
| dtest | [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/17085/) | :white_check_mark: |  |
|  | [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5806/) | :white_check_mark: |  |
|  | [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/9003/) | :white_check_mark: |  |

#### Build Details:
- Duration: 3 hr 33 min

dawmd commented:
Updated the backport section in the cover letter: we no longer try to backport the changes to the previous versions.
